File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
# knapsack 문제 전형적인
N, K = map(int,input().split())

wArr = [0]
cArr = [0]

for i in range(N):
  w,c = map(int,input().split()) # 각 물건의 무게 , 가치 
  wArr.append(w)
  cArr.append(c)

dp = [[0 for i in range(K+1)] for _ in range(N+1)]

for y in range(N+1):
  for x in range(K+1):
    if x < wArr[y]:
      dp[y][x] = dp[y-1][x]
      continue
    #아니라면
    dp[y][x] = max(dp[y-1][x], dp[y-1][x-wArr[y]] + cArr[y])

print(dp[-1][-1])

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.js
const fs = require("fs");
const input = fs.readFileSync('/dev/stdin').toString().split("\n");

const [n, k] = input[0].split(' ').map(Number);
const wArr = [0];
const cArr = [0];

for (let i = 1; i <= n; i++) {
  const [w, v] = input[i].split(' ').map(Number);
  wArr.push(w);
  cArr.push(v);
}

let dp = Array.from({ length: n + 1 }, () => new Array(k + 1).fill(0));

for (let y = 1; y <= n; y++) {
  for (let x = 0; x <= k; x++) {
    if (x < wArr[y]) {
      dp[y][x] = dp[y - 1][x];
      continue;
    }
    dp[y][x] = Math.max(dp[y - 1][x], dp[y - 1][x - wArr[y]] + cArr[y]);
  }
}

console.log(dp[n][k])